### PR TITLE
Format flattened output in kubectl

### LINKED ISSUE #15415: Enabling kubectl get with List makes output funky
Ref https://github.com/kubernetes/kubernetes/pull/14888#issuecomment-147011249
```
➜  kubernetes git:(master) k get po --namespace=kube-system                                 
NAME                             READY     STATUS    RESTARTS   AGE
elasticsearch-logging-v1-eqoex   1/1       Running   0          2h
NAME                             READY     STATUS    RESTARTS   AGE
elasticsearch-logging-v1-vfpyf   1/1       Running   0          2h
NAME                                                    READY     STATUS    RESTARTS   AGE
fluentd-elasticsearch-e2e-test-mikedanese-minion-1rwb   1/1       Running   0          2h
NAME                                                    READY     STATUS    RESTARTS   AGE
fluentd-elasticsearch-e2e-test-mikedanese-minion-ccus   1/1       Running   0          2h
NAME                                                    READY     STATUS    RESTARTS   AGE
fluentd-elasticsearch-e2e-test-mikedanese-minion-ofta   1/1       Running   0          2h
NAME                 READY     STATUS    RESTARTS   AGE
heapster-v10-v20fa   1/1       Running   0          2h
NAME                      READY     STATUS    RESTARTS   AGE
kibana-logging-v1-ht9sx   1/1       Running   1          2h
NAME                READY     STATUS    RESTARTS   AGE
kube-dns-v9-b0ym3   4/4       Running   0          2h
NAME               READY     STATUS    RESTARTS   AGE
kube-ui-v3-v0yrj   1/1       Running   0          2h
NAME                                   READY     STATUS    RESTARTS   AGE
monitoring-influxdb-grafana-v2-yy6u2   2/2       Running   0          2h
```

cc @mikedanese @bgrant0607

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #15415

cc @bgrant0607 @kubernetes/rh-ux @kubernetes/kubectl @ncdc @derekwaynecarr

### CODE REVIEW

**Comment by ncdc** [comment]
I'm not sure this is 100% correct. I have a json file with some service accounts in it. I'm seeing this:

```
vagrant@localhost:~/go/src/k8s.io/kubernetes (JanetKuo-kubectl-get-funky) k get -f sa.json
NAME                SECRETS   AGE
whatever-you-want   0         <unknown>
whatever-you-want-really   0         <unknown>
whatever-you-want-really-long   0         <unknown>
whatever-you-want-really-long-name   0         <unknown>

vagrant@localhost:~/go/src/k8s.io/kubernetes (JanetKuo-kubectl-get-funky) k get serviceaccount
NAME                                 SECRETS   AGE
default                              1         9m
whatever-you-want                    1         6m
whatever-you-want-really             1         4m
whatever-you-want-really-long        1         4m
whatever-you-want-really-long-name   1         4m
```

File: pkg/kubectl/resource_printer.go [hunk 1]
```diff
@@ -1480,6 +1480,28 @@ func (h *HumanReadablePrinter) PrintObj(obj runtime.Object, output io.Writer) er
 	w := tabwriter.NewWriter(output, tabwriterMinWidth, tabwriterWidth, tabwriterPadding, tabwriterPadChar, tabwriterFlags)
 	defer w.Flush()
 	t := reflect.TypeOf(obj)
+	// if obj is an api.List object, we first extract its underlying items and then print them individually
+	if t.String() == "*api.List" {
```
Comment (liggitt) [inline pkg/kubectl/resource_printer.go]: does it need to capture all the items of the same kind in a run and print them together?

**Comment by JanetKuo** [comment]
@ncdc what's inside `sa.json`? api.List, api.ServiceList, or a concatenation of Service?

**Comment by ncdc** [comment]
@JanetKuo 

```json
{
  "kind": "List",
  "apiVersion": "v1",
  "items": [
    {"kind":"ServiceAccount","apiVersion":"v1","metadata":{"name":"whatever-you-want"}},
    {"kind":"ServiceAccount","apiVersion":"v1","metadata":{"name":"whatever-you-want-really"}},
    {"kind":"ServiceAccount","apiVersion":"v1","metadata":{"name":"whatever-you-want-really-long"}},
    {"kind":"ServiceAccount","apiVersion":"v1","metadata":{"name":"whatever-you-want-really-long-name"}}
  ]
}
```

File: pkg/kubectl/resource_printer.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/kubectl/resource_printer.go]: That would be nice.

**Comment by JanetKuo** [comment]
After `Flatten()` is removed, `kubectl get -f` with files of type xxxList no longer gives you some generated fields such as `creationTimestamp`. It's an old bug (that we didn't notice) and unrelated to this change. Not sure how to fix it yet.

**Comment by liggitt** [comment]
it seems like what's missing is grouping like objects in a row, and printing them as a block, so that headers are only printed once, and that columns align for all the objects in the run. as long as we have to do that logic, why not keep the Flatten(), and do that grouping/printing in the Get visitor?

File: pkg/kubectl/resource_printer.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/kubectl/resource_printer.go]: I'm confused why PrintObj is getting passed a List.  It should never do so if the Visitor pattern is being used.

Comment (liggitt) [inline pkg/kubectl/resource_printer.go]: the Flatten() was removed

**Comment by smarterclayton** [comment]
Previously flatten would return a list of objects in the order they were retrieved, which was *always* by type (because we have no API that returns different types of objects in the same rest call).  Why are objects being returned by resource.Builder that aren't grouped by type?  The default behavior of resource builder should be to group returned objects by type (and if you don't want that you'd say GroupByKind(false) or something).

File: pkg/kubectl/resource_printer.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/kubectl/resource_printer.go]: It shouldn't be.

**Comment by smarterclayton** [comment]
Both Flattening and GroupingByKind are properly concerns of the resource.Builder - it should be easy to transform the list of traversals by group even before retrieval.  One challenge is when a user says:

    kubectl describe service/bar rc/foo

That's an operation that probably should not be grouped by type.  I'm not sure whether get should have the same semantics - returning resources in the order the user specified them is fulfilling the user's default expectation.

**Comment by liggitt** [comment]
@smarterclayton I think HumanReadablePrinter has to be updated to handle non-homogenous api.List types correctly

**Comment by liggitt** [comment]
ha, figured it out... keep the Flatten(), and share the printer between visitor invocations as long as the mapping is the same, rather than recreating it on every visit:
```
	// use the default printer for each object
	printer = nil
	var lastMapping *meta.RESTMapping
	return b.Flatten().Do().Visit(func(r *resource.Info, err error) error {
		if err != nil {
			return err
		}
		if printer == nil || lastMapping == nil || r.Mapping == nil || r.Mapping.Resource != lastMapping.Resource {
			printer, err = f.PrinterForMapping(cmd, r.Mapping, allNamespaces)
			if err != nil {
				return err
			}
			lastMapping = r.Mapping
		}
		return printer.PrintObj(r.Object, out)
	})
```

**Comment by liggitt** [comment]
with that change, the HumanReadablePrinter special handling of *api.List should be able to go away

**Comment by JanetKuo** [comment]
Thanks! I'll make that change.

**Comment by smarterclayton** [comment]
Which was what the previous code did

On Oct 12, 2015, at 9:46 PM, Jordan Liggitt <notifications@github.com>
wrote:

ha, figured it out... keep the Flatten(), and share the printer between
visitor invocations as long as the mapping is the same, rather than
recreating it on every visit:

    // use the default printer for each object
    printer = nil
    var lastMapping *meta.RESTMapping
    return b.Flatten().Do().Visit(func(r *resource.Info, err error) error {
        if err != nil {
            return err
        }
        if printer == nil || lastMapping == nil || r.Mapping == nil ||
r.Mapping.Resource != lastMapping.Resource {
            printer, err = f.PrinterForMapping(cmd, r.Mapping, allNamespaces)
            if err != nil {
                return err
            }
            lastMapping = r.Mapping
        }
        return printer.PrintObj(r.Object, out)
    })

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/15499#issuecomment-147571053>
.

**Comment by ncdc** [comment]
@smarterclayton historically, prior to #14888, `Flatten()` was not being called all the time, so when you tried to do `kubectl get -f <file with an api.List>`, it would result in an error. Flatten was added in 14888, which fixed get-file-with-List, but it also resulted in the misaligned columns.

**Comment by smarterclayton** [comment]
So primary question - was an api.List being passed to printObj, and
returned from the visitor?  The visitor shouldn't return an API.List when
Latest() is specified - either Latest must imply Flatten, or something is
broken in traversing files from input.

On Oct 13, 2015, at 8:54 AM, Andy Goldstein <notifications@github.com>
wrote:

@smarterclayton <https://github.com/smarterclayton> historically, prior to
#14888 <https://github.com/kubernetes/kubernetes/pull/14888>, Flatten() was
not being called all the time, so when you tried to do kubectl get -f <file
with an api.List>, it would result in an error. Flatten was added in 14888,
which fixed get-file-with-List, but it also resulted in the misaligned
columns.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/15499#issuecomment-147706314>
.

**Comment by JanetKuo** [comment]
@smarterclayton Before #14888 (which is merged last week), not only api.List, but also other list-type objects such as api.PodList, api.ServiceList are passed to printObj and returned from the visitor. Then in [resource_printer.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L412) we can use the HumanReadablePrinter to handle printing either resource or resourceList. 

Looks like in [builder.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource/builder.go#L692) we don't say Latest must imply Flatten. IIRC, `Latest()` is added when we wanted to support `kubectl get` with file (`-f`).

**Comment by smarterclayton** [comment]
I think this was an oversight in the flow - Flatten() for things returned
from the server, but not handling when inputs are lists and need to be
refreshed.  I *think* this means that kubectl get is broken because it's
not actually going out to the server and refreshing the individual items.
Which means we need the equivalent of flatten to be happening automatically
on input (not requiring it on output).  So we probably need to have the
list object that we load from file have the equivalent of a "flatten
visitor" applied to it prior to getting elsewhere in the code.  That'll
allow the "Latest()" wrapper visitor to refresh the Info.

**Comment by ncdc** [comment]
@JanetKuo are you going to take a shot at implementing what @smarterclayton suggested?

**Comment by JanetKuo** [comment]
@ncdc I just updated the code -- keep `Flatten()` and format the output in get.go.

**Comment by smarterclayton** [comment]
I'm concerned other places are broken - effectively if we go down this path
*every* client path (create / delete / update / replace) should be using
Flatten().  We should verify that's the case.

File: pkg/kubectl/cmd/get.go [hunk 2]
```diff
@@ -220,13 +222,23 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 	}
 
 	// use the default printer for each object
+	printer = nil
+	var lastMapping *meta.RESTMapping
+	w := tabwriter.NewWriter(out, kubectl.TabwriterMinWidth, kubectl.TabwriterWidth, kubectl.TabwriterPadding, kubectl.TabwriterPadChar, kubectl.TabwriterFlags)
```
Comment (liggitt) [inline pkg/kubectl/cmd/get.go]: why is this required?

Comment (JanetKuo) [inline pkg/kubectl/cmd/get.go]: tabwriter translates tabbed columns in input into properly aligned text. Without this, 
```
NAME                                 SECRETS   AGE
default                              1         9m
whatever-you-want                    1         6m
whatever-you-want-really             1         4m
whatever-you-want-really-long        1         4m
whatever-you-want-really-long-name   1         4m
```
becomes something like 
```
NAME    SECRETS   AGE
default 1         9m
whatever-you-want   1         6m
whatever-you-want-really   1         4m
whatever-you-want-really-long   1         4m
whatever-you-want-really-long-name   1         4m
```

Comment (JanetKuo) [inline pkg/kubectl/cmd/get.go]: We make object to be printed with HumanReadablePrinter share the same tabwriter (so that they'll be aligned).

**Comment by JanetKuo** [comment]
@smarterclayton yes, other commands use `Flatten()`. I think the reason that `get` didn't use `Flatten()` is because it's easier to print the objects aligned as a group with HumanReadablePrinter.

**Comment by JanetKuo** [comment]
The command that don't use `Flatten()` are `cluster-info` and `rolling-update`, and they don't take List as input.

File: pkg/kubectl/cmd/get.go (hunk 2 above)
Comment (liggitt) [inline pkg/kubectl/cmd/get.go]: hmm... reusing the writer that way is gorpy in two places... any cleaner way to accomplish that?

**Comment by JanetKuo** [comment]
@smarterclayton According to [resource builder](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource/builder.go#L701), [`RetrieveLatest`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource/visitor.go#L563) should be visited when we specify filepath in `kubectl get` if we use `Lastest()`. I found that if the file contains types like "api.List" or "api.xxxList", `RetrieveLatest` won't be called, and therefore list objects won't be refreshed. (After we add `Flatten()` this problem goes away since list objects are all flattened). 

Another problem is that "watch" only takes a single resource so we can't do flatten. Then do we allow passing a List (or xxxList) to PrintObj()? If not, watch operations like `kubectl get pods -w` won't work anymore.

File: pkg/kubectl/cmd/get.go (hunk 2 above)
Comment (JanetKuo) [inline pkg/kubectl/cmd/get.go]: I'm not sure if there's a better way. We need to do `Flush()` only after visiting all objects. The tabwriter in HumanReadablePrinter.PrintObj() is different; it's for callers other than get.go, so that their output also get aligned.

**Comment by smarterclayton** [comment]
I don't know for sure whether I think RetrieveLatest is broken, or working
as intended.  It's probably ok to require Flatten() today.

I may not understand your question about watch - watch should be a
different enough code path that we don't hit this issue (because we don't
Visit the watch, we invoke the Watch() method).

**Comment by JanetKuo** [comment]
@smarterclayton watch will print the object first (with`PrintObj`) before watch. This means we need to support passing a list-like object to `PrintObj`.

File: pkg/kubectl/cmd/get.go (hunk 2 above)
Comment (JanetKuo) [inline pkg/kubectl/cmd/get.go]: Wrap the creation of tabwriter as a function.

File: pkg/kubectl/resource_printer.go [hunk 3]
```diff
@@ -1475,10 +1475,19 @@ func formatWideHeaders(wide bool, t reflect.Type) []string {
 	return nil
 }
 
+// GetNewTabWriter returns a tabwriter that translates tabbed columns in input into properly aligned text.
+func GetNewTabWriter(output io.Writer) *tabwriter.Writer {
+	return tabwriter.NewWriter(output, tabwriterMinWidth, tabwriterWidth, tabwriterPadding, tabwriterPadChar, tabwriterFlags)
```
Comment (JanetKuo) [inline pkg/kubectl/resource_printer.go]: @liggitt how about wrapping it as a function?

**Comment by ncdc** [comment]
@JanetKuo are we good with watch, or do we still need to make some changes?

**Comment by JanetKuo** [comment]
We're not making changes now. PTAL

File: pkg/kubectl/cmd/get.go [hunk 4]
```diff
@@ -220,13 +221,23 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
...
-			return err
+		if printer == nil || lastMapping == nil || r.Mapping == nil || r.Mapping.Resource != lastMapping.Resource {
+			printer, err = f.PrinterForMapping(cmd, r.Mapping, allNamespaces)
+			if err != nil {
+				return err
+			}
+			lastMapping = r.Mapping
+		}
+		if _, found := printer.(*kubectl.HumanReadablePrinter); found {
+			return printer.PrintObj(r.Object, w)
```
Comment (ncdc) [inline pkg/kubectl/cmd/get.go]: Why do you pass the TabWriter if it's a HumanReadablePrinter vs out if it's not?

Comment (ncdc) [inline pkg/kubectl/cmd/get.go]: In what situations would it not be a HumanReadablePrinter?

Comment (JanetKuo) [inline pkg/kubectl/cmd/get.go]: There are lots of other printers, such as YAMLPrinter, JSONPrinter, etc. See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L61. Only HumanReadablePrinter needs a tabwriter to align its table output.

Comment (ncdc) [inline pkg/kubectl/cmd/get.go]: Cool, thanks!

Comment (JanetKuo) [inline pkg/kubectl/cmd/get.go]: Users use `-o` or `--output` to specify different output format (use different printers).

**Comment by ncdc** [comment]
I don't believe I have any other comments, unless there was some way to write test cases for this? @smarterclayton @liggitt do you have anything else?

**Comment by JanetKuo** [comment]
If there's no objection shall we put LGTM? The related issue is quite urgent.

**Comment by ncdc** [comment]
LGTM as well, thanks!
